server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java

```diff
@@ -142,6 +147,37 @@ public void testCancelException() {
         assertPropagatesInterrupt(() -> future.actionGet(10, TimeUnit.SECONDS));
     }
 
+    public void testAssertCompleteAllowedAllowsConcurrentCompletesFromSamePool() {
+        assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
+        final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
+        final var executorName = randomFrom(ThreadPool.Names.GENERIC, ThreadPool.Names.MANAGEMENT);
+        final var threadCount = 4;
+        final var running = new AtomicBoolean(true);
+        final var startBarrier = new CyclicBarrier(threadCount + 1);
+        final var futures = new ArrayList<Future<?>>();
+        try (TestThreadPool threadPool = new TestThreadPool(getTestName())) {
+            assert threadPool.info(executorName).getMax() >= threadCount : "Not enough threads in pool";
```

Comment:
Why not use all the threads available? More than 4 normally, but even if the pool is smaller than 4 sometimes it's still a valid test.

Reply:
It blows the test time out to 300ms+ on my machine because of the additional synchronisation. 4 threads takes < 100ms but still reliably reproduces the issue.

Reply:
Ok let's use `Math.min(threadPool.info(executorName).getMax(), 4)` rather than failing the test if the threadpool is smaller than 4, and add a comment about why.